sw/device/lib/base/memory.h

```diff
@@ -69,6 +80,38 @@ inline uint32_t read_32(const void *ptr) {
   return val;
 }
 
+/**
+ * Load a 64-bit word from memory directly, bypassing aliasing rules.
+ *
+ * ISO C forbids, in general, casting a pointer to non-character types and
+ * reading them, though it is frequently necessary to read exactly one 64-bit
+ * word out of a `void *`. This function performs that action in a manner which
+ * is well-defined.
+ *
+ * Of course, `ptr` must point to word-aligned memory that is at least one
+ * 64-bit word wide. To do otherwise is Undefined Behavior. It goes eithout
+ * saying that the memory this function intents to read must be initialized.
+ *
+ * This function has reordering properties as weak as a normal, non-atomic,
+ * non-volatile load.
+ *
+ * @param ptr a word-aligned pointer pointed to at least four bytes of memory.
+ * @return the word `ptr` points to.
+ */
+static inline uint64_t read_64(const void *ptr) {
+  // Both GCC and Clang optimize the code below into a single word-load on most
+  // platforms. It is necessary and sufficient to indicate to the compiler that
+  // the pointer points to four bytes of four-byte-aligned memory.
+  //
+  // Failing to get that particular codegen in either GCC or Clang with -O2 or
+  // -Os set shall be considred a bug in this function. The same applies to
+  // `write32()`.
+  ptr = __builtin_assume_aligned(ptr, alignof(uint32_t));
```

Comment:
I think you got the alignof wrong here.

Reply:
Interesting.  On RV32I, where uint64_t is a manipulated as two 32-bit words, 8-byte alignment shouldn't necessarily be required (but the compiler _does_ align 64-bit words as such in structs).

However, the `{read, write}_64` functions, as implemented, generate correct code:

```
static uint8_t dequeue_byte(queue_t *queue) {
  uint8_t val = queue->data[0];
  uint64_t qword = read_64(queue->data);
        lw a1,8(sp)
        lw a2,4(sp)

  // assembly for the shift omitted.
  write_64(qword >> 8, queue->data);
        sw a1,8(sp)
        sw a2,4(sp)

  queue->length -= 1;
  return val;
}
```

If I change the alignment constraint, I'll have to force `struct queue` to have 8-byte alignment for the `data` field.  When it was implemented as a union of `uint` types, the compiler just handled it automatically for me.

Reply:
After offline discussion, I'm going to align to `uint64_t` as that is the architectural constraint for rv32i:
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#user-content-cc-type-details